### PR TITLE
Feedforward neural network RLearner for classification using MXNet

### PR DESCRIPTION
**Pull request by bklppr**
Implementation of an `RLearner` for creating a feedforward neural network with up to three layers using the `mxnet` package. The learner makes use of `mxnet`'s symbols and is easily extendable for enhancing its functionalities in the future.
See also issue #1120 .

### CODE REVIEW

**Comment by larskotthoff** [comment]
Thanks. It looks like the listLearners tests are failing now; could you have a look please? Also please double check the style guide, there's at least one <- for assignment.

**Comment by bklppr** [comment]
Thanks for the feedback, both things should be fixed now.

**Comment by larskotthoff** [comment]
Thanks. Could you also fix the lintr errors please?

File: R/RLearner_classif_mxff.R [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+}
+
+#' @export
+trainLearner.classif.mxff = function(.learner, .task, .subset, .weights = NULL,
+  layers = 1L, nodes1 = 1L, nodes2 = NULL, nodes3 = NULL, nodes.out = NULL,
+  act1 = "tanh", act2 = NULL, act3 = NULL, act_out = "softmax", dropout = NULL, symbol = NULL,
+  ...) {
+  # transform data in correct format
+  d = getTaskData(.task, subset = .subset, target.extra = TRUE)
+  y = as.numeric(d$target) - 1
```
Comment (larskotthoff) [inline R/RLearner_classif_mxff.R]: Why the -1 here? Will this always work?

File: R/RLearner_classif_mxff.R [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+    # construct hidden layers using symbols
+    for (i in seq_len(layers)) {
+      sym = mx.symbol.FullyConnected(sym, num_hidden = nodes[[i]])
+      sym = mx.symbol.Activation(sym, act_type = act[[i]])
+    }
+
+    # add dropout if specified
+    if (!is.null(dropout)) {
+      dropout = min(dropout, (1 - 1e-7))
```
Comment (larskotthoff) [inline R/RLearner_classif_mxff.R]: Why is dropout adjusted here? Should this not be in the parameter definition?

File: R/RLearner_classif_mxff.R (hunk 1 above)
Comment (bklppr) [inline R/RLearner_classif_mxff.R]: The encoding for `mx.model.FeedForward.create()` needs to start at 0.
As `d$target` is a factor (`.task` is a `ClassifTask`), `as.numeric(d$target)` will create a numeric encoding starting at 1, hence the -1.

File: R/RLearner_classif_mxff.R (hunk 2 above)
Comment (bklppr) [inline R/RLearner_classif_mxff.R]: `dropout` needs to be in [0,1) and the `mxnet` function `mx.mlp()` which also implements this had it like that.
But you're absolutely right, this should be made explicit. I changed the upper bound for `dropout` and also commented on it in the notes.

**Comment by larskotthoff** [comment]
Thanks. It looks like you've changed the file in todo-files as well -- that can be deleted, can't it?

**Comment by bklppr** [comment]
Yes you´re right. Thanks for your help!

**Comment by larskotthoff** [comment]
Great, thanks. Merging!

**Comment by rwarnung** [comment]
@bklppr Hi, thank you for the work on this learner. If I see it correctly then the prediction of probabilities is not supported in the learner. But it is supported in the package with `out_activation     = "softmax"`. Did you leave this out on purpose? Thanks again and best regards!

**Comment by bklppr** [comment]
@rwarnung Hi, thanks for your nice words!
I did not leave it out on purpose, I was planning to add that feature soon, it should not be too complicated.
Of course you are welcome to extend the learner yourself if it is very urgent :)
Feel free to consult me in this case if needed. Otherwise I will add it soon.
Best regards!

**Comment by rwarnung** [comment]
Thank your for your reply and the offer to help. I might find time in
summer (end of June) for the probabilities. Let's see who's quicker :),
again thanks for your work on this. best regards, Richard

2017-06-02 17:13 GMT+02:00 Benjamin Klepper <notifications@github.com>:

**Comment by bklppr** [comment]
@rwarnung I added the feature yesterday ;) Please see #1826
Best regards!

**Comment by rwarnung** [comment]
That's great!!! I will check it out next week. Really great! Thank you!

Am 05.06.2017 9:08 vorm. schrieb "Benjamin Klepper" <
notifications@github.com>:
